### PR TITLE
De-jsonificate logins crate

### LINKED ISSUE #1605: The logins crate should use protobufs
The following methods return their results as JSON because we `implement_into_ffi_by_json!(Login);`
https://github.com/mozilla/application-services/blob/fe9c866787898d40fae1294d218a211c0b8c0c15/components/logins/ffi/src/lib.rs#L166-L201

The work to do is [the same as what we've done on the Push crate](https://github.com/mozilla/application-services/issues/1491#issue-474206810), plus:

* Swift support ([example](https://github.com/mozilla/application-services/blob/4885cd11fcbacebaa707f7d78878ca15918b7c38/components/places/ios/Places/Places.swift#L331-L345))
* (Optional/Good follow-up) `sync15_passwords_add` also takes JSON as input! What if we could take protobuf as an input instead? We do this for the places crate in [Kotlin](https://github.com/mozilla/application-services/blob/4885cd11fcbacebaa707f7d78878ca15918b7c38/components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt#L488-L495) and [Swift](https://github.com/mozilla/application-services/blob/4885cd11fcbacebaa707f7d78878ca15918b7c38/components/places/ios/Places/Places.swift#L777-L787).
* Finally `implement_into_ffi_by_json!(Login);` can be deleted.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-743)

### PR DESCRIPTION
**Pull request by lyuyuan**
fixes #1605 

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `automation/all_tests.sh` runs to completion and produces no failures
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

**Comment by lyuyuan** [comment]
The crashed test is in FxAccountManagerTests. It somehow doesn't crash on my local build.

File: components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt [hunk 1]
```diff
@@ -366,6 +372,19 @@ internal fun Pointer.getAndConsumeRustString(): String {
     }
 }
 
+/**
+ * Helper to read a protocal buffer.
+ *
+ * Important: Do not use this buffer after this!
+ */
+internal fun RustBuffer.ByValue.getAndConsumeBuffer(): CodedInputStream? {
```
Comment (eoger) [inline components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt]: Can we inline this in the functions above, so we can wrap the `parseFrom` calls in the `try`?

File: components/logins/ios/Logins/LoginRecord.swift [hunk 2]
```diff
@@ -171,3 +171,26 @@ open class LoginRecord {
...
     }
 }
+
+internal func unpackProtobufInfo(msg: MsgTypes_PasswordInfo) -> LoginRecord {
+    return LoginRecord(
+        id: msg.id,
+        password: msg.password,
+        hostname: msg.hostname,
+        username: msg.username,
+        formSubmitURL: msg.formSubmitURL.isEmpty ? nil : msg.formSubmitURL,
```
Comment (eoger) [inline components/logins/ios/Logins/LoginRecord.swift]: I think you want the `has<Property>` method pattern in Swift too.

**Comment by lyuyuan** [comment]
Hi @eoger let me know if I missed any requested changes.

> We should also make sure to smoke test this against a real application such a Lockwise before landing to ensure everything works as usual :) I can take this or can help you set it up if you are interested in end-to-end testing your changes.

Sure! Could you give me some guidance on how to do this?

File: components/logins/ios/Logins/LoginsStorage.swift [hunk 3]
```diff
@@ -338,38 +340,42 @@ open class LoginsStorage {
     open func get(id: String) throws -> LoginRecord? {
         return try queue.sync {
             let engine = try self.getUnlocked()
-            let ptr = try LoginsStoreError.tryUnwrap { err in
+            let buffer = try LoginsStoreError.tryUnwrap { err in
```
[the remark this answers is not in the document]
Comment (eoger) [reply components/logins/ios/Logins/LoginsStorage.swift]: To make this clear about this change and the one above: we always return a `RustBuffer` no matter, however the `data` field can be null to signal this is a `None` mapping.

File: components/logins/src/login.rs [hunk 4]
```diff
@@ -543,6 +544,37 @@ impl Login {
...
+            username_field: if login.username_field.is_empty() {
+                None
+            } else {
+                Some(login.username_field)
+            },
+            password_field: if login.password_field.is_empty() {
+                None
+            } else {
+                Some(login.password_field)
+            },
```
[the remark this answers is not in the document]
Comment (lyuyuan) [reply components/logins/src/login.rs]: I think [this](https://github.com/mozilla/application-services/commit/617fef2fd623c55066e2377b6dc3ef127e211185#diff-3c9f89a875af7d9d5c12376d169a85dd) was the reason why I wrote them like these several months ago. I agree only `form_submit_url` and `http_realm` would be nullable now.

File: components/logins/ios/Logins/LoginsStorage.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (lyuyuan) [reply components/logins/ios/Logins/LoginsStorage.swift]: I see. Thanks for clarifying this for me.

**Pull request merged**